projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,550 @@
+    return result[:min(10, len(result))]
+  
+print "Show only 10 info lines per problem \n"
+for err in errorsPerProblem.keys():
+    if err != '':
+        print err
+        print "\n".join(map(printOSReason, osAndReasonForErros(err)))
+        print "\n"
+```
+    Show only 10 info lines per problem 
+    
+    sendFailure count must not be more than 60
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'6.1', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'3.10.65', u'name': u'Android'} reason: delayed
+    os: {u'version': u'4.9.0-kali1-amd64', u'name': u'Linux'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    
+    
+
+
+# Nightly 56 & 57. Compute pings count per day
+This include showing diagrams and printings stats
+
+
+```python
+from datetime import datetime
+
+def pingsCountPerDay(pings):
+    return pings.map(lambda ping: ping["meta/submissionDate"]).countByValue()
+
+resultDictionary = pingsCountPerDay(cachedData)
+```
+
+```python
+import matplotlib.dates as mdates
+def plotlistofTuples(listOfTuples, title="", inColor='blue'):
+    keys = [t[0] for t in listOfTuples]
+    values = [t[1] for t in listOfTuples]
+
+    plt.figure(1)
+    fig = plt.gcf()
+    fig.set_size_inches(15, 7)
+
+    plt.title(title)
+    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter('%m/%d/%Y'))
+    plt.bar(range(len(listOfTuples)), values, align='center', color=inColor)
+    plt.xticks(range(len(listOfTuples)), keys, rotation=90)
+
+```
+
+```python
+plotlistofTuples(sorted(resultDictionary.items(), key=lambda tup: tup[0]), "Pings count per day")  
+```
+
+
+![png](images/output_23_0.png)
+
+
+
+```python
+pingsPerDaySeries = pd.Series(resultDictionary.values())
+pingsPerDaySeries.describe([.25, .5, .75, .95])
+```
+
+
+
+
+    count        27.000000
+    mean      74711.481481
+    std       37555.797027
+    min        6475.000000
+    25%       54222.500000
+    50%       70895.000000
+    75%       82903.000000
+    95%      138705.400000
+    max      164572.000000
+    dtype: float64
+
+
+
+# Nightly 56 & 57. # Compute how many clients are reporting
+
+
+```python
+def getClients(pings):
+    clients = pings.map(lambda ping: ping["clientId"]).distinct()
+    return clients
+```
+
+```python
+clientsCount = getClients(cachedData).count()
+print "Clients number = " + str(clientsCount)
+```
+    Clients number = 120336
+
+
+# Nightly 56 & 57. Compute average number of pings per client per day
+
+* Expected ping per day = 24, as we send no one Health ping per hour. 
```

Comment:
"We expect at most 24 pings per day as we send no more than one "health" ping per hour"